### PR TITLE
feat: bootstrap the registry resource specs

### PR DESCRIPTION
**Pull request by paulyoong**
Make the registry initialise the resource specs with the contents of the
persistent store. This ensures that the specs in the registry match what
was there prior to the core agent being restarted. This makes the core
agent more resilient across restarts.

The resource specs are also now exposed through the REST API. This
allows the caller to inspect the requested configuration of the various
resources. If any resource is unhealthy, the spec will be the
configuration that the reconciler will attempt to achieve.

Additional changes:
- Move type definitions to their own directory to avoid cyclical
dependencies.

Resolves: CAS-827

### CODE REVIEW

**Comment by paulyoong** [comment]
bors try

File: tests-mayastor/tests/nexus.rs [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
+use types::v0::message_bus::mbus;
 
 #[actix_rt::test]
 async fn create_nexus_malloc() {
     let cluster = ClusterBuilder::builder().build().await.unwrap();
 
     cluster
         .rest_v0()
-        .create_nexus(v0::CreateNexus {
+        .create_nexus(mbus::CreateNexus {
```
Comment (tiagolobocastro) [inline tests-mayastor/tests/nexus.rs:11]: All of the tests which use the rest client should use the rest types (v0) and not the mbus types as they are not necessarily the same - it's an implementation detail.

File: control-plane/rest/src/versions/v0.rs [hunk 2]
```diff
@@ -12,6 +12,16 @@ use std::{
     string::ToString,
 };
 use strum_macros::{self, Display};
+use types::v0::message_bus::mbus::{
```
Comment (tiagolobocastro) [inline control-plane/rest/src/versions/v0.rs:15]: this should be pub use to export the types which are part of the rest client interface. The user of the rest client interface should only require to use the rest_client library - in theory :)

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/store/src/etcd.rs [hunk 3]
```diff
@@ -115,6 +115,24 @@ impl Store for Etcd {
         }
     }
 
+    /// Retrieve objects with the given key prefix
+    async fn get_values_prefix(&mut self, key_prefix: String) -> Vec<(String, Value)> {
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/store/src/etcd.rs]: Done.

File: control-plane/store/src/etcd.rs [hunk 4]
```diff
@@ -115,6 +115,24 @@ impl Store for Etcd {
...
+        let resp = self
+            .0
+            .get(key_prefix, Some(GetOptions::new().with_prefix()))
+            .await
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/store/src/etcd.rs]: Done.

File: control-plane/rest/src/versions/v0.rs [hunk 5]
```diff
@@ -521,6 +528,11 @@ impl RestClient for ActixRestClient {
         );
         self.del(urn).await
     }
+
+    async fn get_specs(&self) -> ClientResult<Specs> {
+        let urn = "/v0/private/specs".to_string();
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/rest/src/versions/v0.rs]: Done.

File: control-plane/agents/core/src/core/specs.rs [hunk 6]
```diff
@@ -33,6 +40,84 @@ impl ResourceSpecsLocked {
...
+
+    /// Populate the resource specs with data from the persistent store.
+    async fn populate_specs<S: Store>(&self, store: &mut S, spec_type: StorableObjectType) {
+        let prefix = key_prefix(spec_type);
+        let store_specs = store.get_values_prefix(prefix).await;
+        let mut resource_specs = self.0.write().await;
+
+        for (key, value) in store_specs {
+            // The uuid is assumed to be the last part of the key.
+            let id = key.split('/').last().expect("Failed to get uuid from key");
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/agents/core/src/core/specs.rs]: Done. Returning an error now.

File: control-plane/agents/core/src/core/specs.rs [hunk 7]
```diff
@@ -33,6 +40,84 @@ impl ResourceSpecsLocked {
...
+            StorableObjectType::PoolSpec,
+            StorableObjectType::ReplicaSpec,
+        ];
+        for spec in &spec_types {
+            self.populate_specs(store, *spec).await;
+        }
+    }
+
+    /// Populate the resource specs with data from the persistent store.
+    async fn populate_specs<S: Store>(&self, store: &mut S, spec_type: StorableObjectType) {
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/agents/core/src/core/specs.rs]: This part is only to load the specs from the persistent store. The committing/clearing of the operation will be left up to the reconciler to figure out.

File: control-plane/agents/core/src/core/tests.rs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    // Restart the core agent with the expectation that the registry will have all its resource
+    // specs loaded from the persistent store.
+    cluster
+        .composer()
+        .restart("core")
+        .await
+        .expect("Failed to restart core agent");
+
+    // Give some time for the core agent to restart.
+    sleep(std::time::Duration::from_secs(2));
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/agents/core/src/core/tests.rs]: Thanks, done.

File: tests-mayastor/tests/nexus.rs (hunk 1 above)
Comment (paulyoong) [reply tests-mayastor/tests/nexus.rs]: Done.

File: control-plane/rest/src/versions/v0.rs (hunk 2 above)
Comment (paulyoong) [reply control-plane/rest/src/versions/v0.rs]: Done.

File: control-plane/rest/src/lib.rs [hunk 9]
```diff
@@ -98,24 +99,38 @@ impl ActixRestClient {
...
             details: format!("Failed to get_vec uri {}", uri),
         })?;
-
         Self::rest_vec_result(rest_response).await
     }
+
+    async fn do_get(
+        &self,
+        uri: &str,
+    ) -> Result<ClientResponse<Decoder<Payload<PayloadStream>>>, SendRequestError> {
```
[the remark this answers is not in the document]
Comment (paulyoong) [reply control-plane/rest/src/lib.rs:126]: As discussed, this wasn't as straightforward as it originally looked. Leaving it as it is for the time being.

**Comment by paulyoong** [comment]
bors merge

**Comment by paulyoong** [comment]
bors merge

**Pull request merged**